fix(components): [tree] dragging a node will deselect the node

Alixhan opened the pull request:
Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

## Description

同时开启多选框和拖拽功能后，拖拽节点会导致节点取消选中

## Related Issue

Fixes #13912.

## Explanation of Changes

拖拽节点时插入新位置时，将checked状态同步带过去即可

Alixhan committed:
fix(components): [tree] Dragging a node will deselect the node(#13912)

btea reviewed (approved):
LGTM

boomboy4 commented:
@btea I've tried this before, but it causes the parent and child nodes before and after the move to have incorrect checkboxes.
I don't think that's right.

boomboy4 commented:
![Uploading 1700620772757.png…]()
Like this

Alixhan commented:
@boomboy4 I just found another bug, unrelated to this PR, which causes the checkbox state of the parent node of the changing node to not be updated, Let me fix it.

Alixhan commented:
Are you talking about that?

boomboy4 commented:
> Are you talking about that?
I mean you only change the state of the dragging node, but you don't manage the old parent-child and new parent-child states of the dragging node

boomboy4 commented:
> Are you talking about that?

I mean you only change the state of the dragging node, but you don't manage the old parent-child and new parent-child states of the dragging node

boomboy4 commented:
<img width="303" alt="a6a49aed7782e4ee3d7d590a1c8db04" src="https://github.com/element-plus/element-plus/assets/37095891/31bd3a62-2be9-44a6-8aa2-6242a956fc64">
Before drag

boomboy4 commented:
<img width="336" alt="1ade5ce04a2df614b07226c823af724" src="https://github.com/element-plus/element-plus/assets/37095891/4bd9adee-1471-4750-9549-340d23637e92">
dragged

Alixhan commented:
@boomboy4 I know your mean，but it is another bug，It existed before I submitted this PR and I just found out about it.

Alixhan commented:
@boomboy4 I'm trying to fix it

boomboy4 commented:
> @boomboy4 I'm trying to fix it

ok. the pr supposed to solve the problem of state after dragging and dropping

Alixhan committed:
fix(components): [tree] Dragging a node will deselect the node (#13912)

Alixhan commented:
@boomboy4 @btea I added a listener to  `tree-node` component to watch for changes to the childNode's length，It can also solve the problem of [#14915](https://github.com/element-plus/element-plus/pull/14915)

boomboy4 commented:
> @boomboy4 @btea I added a listener to `tree-node` component to watch for changes to the childNode's length，It can also solve the problem of [#14915](https://github.com/element-plus/element-plus/pull/14915)

This doesn't solve the problem because the listen here only changes the current node, the state above the parent doesn't change.
I don't have permission to review the code though, that's @btea  job. But thanks for your efforts!

Alixhan commented:
@boomboy4 What you may not have noticed is that changing the checked state triggers the checked listener event, which automatically triggers an update to the parent state, I've tested it and it works fine.

Alixhan commented:
Before
![image](https://github.com/element-plus/element-plus/assets/22884361/f1f1eb21-9b02-4930-81a7-4015a4a8307f)
dragged
![image](https://github.com/element-plus/element-plus/assets/22884361/444606ea-cb33-4cd7-b685-67adfced382d)

boomboy4 commented:
> @boomboy4 What you may not have noticed is that changing the checked state triggers the checked listener event, which automatically triggers an update to the parent state, I've tested it and it works fine.

But darggingNode parent's parent, which has no change in the number of children, will not respond.
You can try dragging to deeper levels, the top level won't change state

Alixhan commented:
@boomboy4 Yes, I noticed. This line("tree-node.vue:332")  solves that problem.

boomboy4 commented:
> @boomboy4 Yes, I noticed. This line("tree-node.vue:332") solves that problem.

Yes, but setChecked is limited to checked or unchecked, and node.indeterminate changes do not affect the parent's indeterminate. 
<img width="588" alt="1700723047726" src="https://github.com/element-plus/element-plus/assets/37095891/2b64b9bd-fb9a-4f5e-9c85-0b034797764c">

boomboy4 commented:
> @boomboy4 Yes, I noticed. This line("tree-node.vue:332") solves that problem.

<img width="491" alt="77daa60a5d2ba1801b3450ba6388bf1" src="https://github.com/element-plus/element-plus/assets/37095891/4066edbf-4722-4fc2-ade0-b13dc569f6fb">

<img width="381" alt="b00c7cf76627b68da7cb3d2edd6ce60" src="https://github.com/element-plus/element-plus/assets/37095891/a35bc792-a2f4-48bd-bbfd-94d65d7729d6">

Alixhan committed:
fix(components): [tree] Dragging a node will deselect the node (#13912)

The Node class adds a method to set the checked state, which only affects the parent node state.

Alixhan commented:
@boomboy4 Thank you for recognizing that I have come to understand the importance of being 100% meticulous in resolving issues. HA HA HA^^^

Alixhan committed:
refactor: optimized code

Nullaha commented:
> @boomboy4 Thank you for recognizing that I have come to understand the importance of being 100% meticulous in resolving issues. HA HA HA^^^

我看了一下你的提交，确实解决了https://github.com/element-plus/element-plus/pull/14915

Alixhan committed:
fix(components): [tree] Dragging  a node will deselect the node (#13912)

checkStrictly is true no need reInitCheckek

Alixhan commented:
@yujinpan @ryuhangyeong 为啥不给审批呢？

btea commented:
@Alixhan Hello, the project is driven by the community, and team members all have their main work to do. Recently, it may be that due to busy work, there is less time to maintain the project. 
Therefore, some PRs may be missed or may not receive a response for a long time. Please wait patiently. ❤️

btea commented:
@Alixhan Could you add a test case?

Alixhan commented:
> @Alixhan Could you add a test case?

节点的拖动无法模拟，并且这个是bug修改，前面功能就没有测试用例，有点无从下手。

btea commented:
In the scenario provided by #15477, dragging a node seems to cause an error.

Alixhan commented:
> In the scenario provided by #15477, dragging a node seems to cause an error.
此PR正是解决了这个问题。

btea commented:
For the scenario provided by #15477, the content you modify in this PR will cause an error.

![play](https://github.com/element-plus/element-plus/assets/24516654/56b62d93-db67-4028-a07e-34079f5e6e98)

Alixhan commented:
> For the scenario provided by #15477, the content you modify in this PR will cause an error.

我看了一下，这个tree组件设计之初就是在拖动后，节点对象会被浅复制，并且丢失选中状态，节点的id也会重新生成，要保持拖动后节点的选中状态，必须设置`node-key`属性，否则无法找到拖动后复制的节点设置选中状态.

Alixhan commented:
@btea 抽空合并一下这个PR呗，着急用。

btea commented:
> @btea 抽空合并一下这个PR呗，着急用。

合并发版可能没那么快，急用的话，建议先自行patch。

kooriookami commented:
> @kooriookami Do you think there are any other problems with this? If not, I think it can be merged.

To be honest, I'm not sure if it'll cause any other problems :D

btea commented on packages/components/tree/src/model/useDragNode.ts:
```diff
@@ -180,6 +180,13 @@ export function useDragNodeHandler({ props, ctx, el$, dropIndicator$, store }) {
         store.value.registerNode(draggingNodeCopy)
+        //restore checkbox state after dragging
+        draggingNode.node.eachNode((node) => {
+          store.value.nodesMap[node.data[store.value.key]]?.setChecked(
```
Maybe we should be compatible with the case where `store.node.key` is `undefined`.

Alixhan replied:
可以这样，当store.node.key为undefined，不执行多选框的恢复操作。

Alixhan replied:
我来修改一下

Alixhan committed:
fix(components): [tree] Dragging a node will deselect the node (#13912)

btea reviewed (approved):
@kooriookami  There is no problem so far, we can merge it first.
